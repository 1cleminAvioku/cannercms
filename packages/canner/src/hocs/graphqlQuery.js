import * as React from 'react';
import gql from 'graphql-tag';
import {Query} from 'react-apollo';

export default function withQuery(Com) {
  return class ComWithQuery extends React.Component {
    render() {
      const {graphql, ...restProps} = this.props;
      return (
        <Query query={gql`${graphql}`} >
          {({loading, error, data}) => {
            if (loading) return null;
            if (error) return `Error!: ${error}`;
            const key = Object.keys(data)[0];
            return (
              <Com value={data[key].map(d => ({...d}))} {...restProps} />
            );
          }}
        </Query>
      );
    }
  }
}