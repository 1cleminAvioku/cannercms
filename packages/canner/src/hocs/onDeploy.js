// @flow

import * as React from 'react';
import {get, isArray, isPlainObject, set} from 'lodash';
import RefId from 'canner-ref-id';
import type {HOCProps} from './types';

export default function withOndeploy(Com: React.ComponentType<*>) {
  return class ComponentWithOnDeploy extends React.Component<HOCProps> {
    key: string;
    id: ?string;

    constructor(props: HOCProps) {
      super(props);
      const {pattern, refId, rootValue} = props;
      const {key, id} = splitRefId({
        refId,
        rootValue,
        pattern
      });
      this.key = key;
      this.id = id;
    }

    removeOnDeploy = (arg1: string, callbackId: string) => {
      const {removeOnDeploy} = this.props;
      if (callbackId) {
        return removeOnDeploy(arg1, callbackId);
      } else {
        return removeOnDeploy(this.key, arg1);
      }
    }

    onDeploy = (arg1: string | Function, callback: Function): string => {
      const {onDeploy, refId} = this.props;
      if (typeof arg1 === 'string') {
        return onDeploy(arg1, callback);
      } else {
        // first arguments is a function
        return onDeploy(this.key, result => {
          let restPathArr = refId.getPathArr();
          // if (this.id) {
          //   restPathArr = restPathArr.slice(2);
          // } else {
            restPathArr = restPathArr.slice(1);
          // }
          const {paths, value} = getValueAndPaths(result.data, restPathArr);
          return {
            ...result,
            // $FlowFixMe
            data: set(result.data, paths, arg1(value))
          }
        });
      }
    }

    render() {
      return <Com {...this.props}
        onDeploy={this.onDeploy}
        removeOnDeploy={this.removeOnDeploy}
      />
  }
  };
}


export function splitRefId({
  refId,
  rootValue,
  pattern
}: {
  refId: RefId,
  rootValue: any,
  pattern: string
}) {
  const [key, index] = refId.getPathArr();
  let id;
  if (pattern.startsWith('array')) {
    id = get(rootValue, [key, index, 'id']);
  }
  return {
    key,
    id
  }
}

export function getValueAndPaths(value: Map<string, *>, idPathArr: Array<string>) {
  return idPathArr.reduce((result: any, key: string) => {
    let v = result.value;
    let paths = result.paths;
    if (isPlainObject(v)) {
      if ('edges' in v && 'pageInfo' in v) {
        v = get(v, ['edges', key, 'node']);
        paths = paths.concat(['edges', key, 'node']);
      } else {
        v = get(v, key);
        paths = paths.concat(key);
      }
    } else if (isArray(v)) {
      v = get(v, key);
      paths = paths.concat(key);
    }
    return {
      value: v,
      paths
    }
  }, {
    value,
    paths: []
  });
}