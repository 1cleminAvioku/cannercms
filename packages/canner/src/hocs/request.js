// @flow

import * as React from 'react';
import {generateAction} from '../action';
import {createEmptyData} from 'canner-helpers';
import RefId from 'canner-ref-id';
import type {HOCProps} from './types';
import {update, merge, isArray, isPlainObject, mapValues} from 'lodash';

type changeQueue = Array<{RefId: RefId | {firstRefId: RefId, secondRefId: RefId}, type: any, value: any}>;

export default function withRequest(Com: React.ComponentType<*>) {
  // this hoc will update data;
  return class ComponentWithRequest extends React.Component<HOCProps> {
    onChange = (refId: RefId | {firstRefId: RefId, secondRefId: RefId} | changeQueue, type: any, delta: any, config: any): Promise<*> => {
      let id;
      if (isArray(refId)) { // changeQueue
        const changeQueue = refId;
        // $FlowFixMe
        return Promise.all(changeQueue.map(args => {
          const {refId, type, value} = args;
          return this.onChange(refId, type, value);
        }));
      } else if (refId instanceof RefId) {
        id = refId.toString();
      } else {
        id = {
          // $FlowFixMe: refId should be object here
          firstId: refId.firstRefId.toString(),
          // $FlowFixMe: refId should be object here
          secondId: refId.secondRefId.toString()
        };
      } 
      const {schema, rootValue, request} = this.props;
      const itemSchema = findSchemaByRefId(schema, refId);
      const {relation, items, pattern} = itemSchema;
      const action = createAction({
        relation,
        id,
        type,
        value: delta,
        config,
        rootValue,
        items,
        pattern
      });
      if (!action) {
        throw new Error('invalid change');
      }
      if (action.type === 'NOOP') {
        return Promise.resolve();
      }
      // $FlowFixMe
      return request(action);
    }

    render() {
      return <Com {...this.props} onChange={this.onChange} />;
    }
  };
}

export function createAction({
  relation,
  id,
  type,
  value,
  config,
  rootValue,
  items,
}: { 
  relation: any,
  id: {firstId: string, secondId: string} | string,
  type: 'create' | 'update' | 'delete' | 'swap',
  value: any,
  config: any,
  rootValue: any,
  value: any,
  items: any
}) {
  if (type === 'create') {
    if (!config) {
      const emptyData = createEmptyData(items);
      if (isPlainObject(emptyData)) {
        value = merge(emptyData, value);
      } else {
        value = emptyData;
      }
    }
    if (typeof id === 'string' && id.split('/').length === 1) {
      update(value, 'id', id => id || randomId());
    }
  }
  value = addTypename(value)
  return generateAction({
    id,
    updateType: type,
    value,
    rootValue,
    relation
  });
}

function randomId() {
  return Math.random().toString(36).substr(2, 12);
}

function addTypename(value) {
  if (isPlainObject(value)) {
    update(value, '__typename', typename => typename || null);
    value = mapValues(value, child => addTypename(child));
  } else if (isArray(value)) {
    value = value.map(child => addTypename(child));
  }
  return value;
}

export function findSchemaByRefId(schema: Object, refId: any) {
  let paths = [];
  if (isPlainObject(refId)) {
    paths = refId.firstRefId.getPathArr();
  } else {
    paths = refId.getPathArr();
  }
  let pattern = '';
  return loop(schema, paths, pattern);
}

export function loop(schema: Object, paths: Array<string>, pattern: string): any {
  if (paths.length === 0) {
    return {
      ...schema,
      pattern: removeFirstSlash(pattern)
    };
  }

  if (!schema.type) { // in items of object
    return loop(schema[paths[0]], paths.slice(1), `${pattern}/${schema[paths[0]].type}`);
  }

  if (schema.type === 'json') {
    return {};
  }

  if (schema.type === 'array') {
    if (paths.length === 1) {
      // paths = [index], so just return
      return loop(schema, [], pattern);
    } else if (schema.items.type === 'object') {
      // path[0] is index, so we take the paths[1]
      return loop(schema.items.items[paths[1]], paths.slice(2), `${pattern}/${schema.items.items[paths[1]].type}`);
    }
  }
  return loop(schema.items[paths[0]], paths.slice(1), `${pattern}/${schema.items[paths[0]].type}`);
}

function removeFirstSlash(pattern: string) {
  return pattern.split('/').slice(1).join('/');
}