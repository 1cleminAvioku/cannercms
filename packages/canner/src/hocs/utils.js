// @flow
import {mapValues, get, isPlainObject, isArray} from 'lodash';

export function getValue(value: any, idPathArr: Array<string>) {
  return idPathArr.reduce((result: any, key: string) => {
    if (isPlainObject(result)) {
      if ('edges' in result && 'pageInfo' in result) {
        return get(result, ['edges', key, 'node']);
      }
      return get(result, key);
    } else if (isArray(result)) {
      return get(result, key);
    } else {
      return result;
    }
  }, value);
}

export function parseConnectionToNormal(value: any) {
  if (isPlainObject(value)) {
    if (value.edges && value.pageInfo) {
      return value.edges.map(edge => parseConnectionToNormal(edge.node));
    }
    return mapValues(value, item => parseConnectionToNormal(item));
  } else if (isArray(value)) {
    return value.map(item => parseConnectionToNormal(item))
  } else {
    return value;
  }
}

export function defaultValue(type: string, relation: any) {
  switch (type) {
    case 'connection': {
      return {
        edges: [],
        pageInfo: {
          hasNextPage: false,
          hasPreviousPage: false
        }
      }
    }
    case 'array': {
      return [];
    }
    case 'object': {
      return {};
    }
    case 'boolean': {
      return false;
    }
    case 'number': {
      return 0;
    }
    case 'string': {
      return '';
    }
    case 'relation': {
      if (relation.type === 'toMany') {
        return {
          edges: [],
          pageInfo: {
            hasNextPage: false,
            hasPreviousPage: false
          }
        };
      } else {
        return null;
      }
    }
    case 'image':
    case 'file': {
      return {
        url: '',
        contentType: '',
        name: '',
        size: 0,
        __typename: null
      }
    }
    case 'geoPoint': {
      return {
        placeId: '',
        address: '',
        lat: 122,
        lng: 23
      };
    }
    default: {
      return null;
    }
  }
}