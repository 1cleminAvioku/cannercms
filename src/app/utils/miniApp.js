// @flow
// miniApp is a proxy to cache data
// used in popup

import App from '../app';
import {CacheStore, Bucket, MiniEndpointMiddleware} from '../middleware';

export default class MiniApp {
  app: App;
  constructor({request, fetch, subscribe}: {request: Function, fetch: Function, subscribe: Function}) {
    this.app = new App()
      .use(new CacheStore())
      .use(new Bucket())
      .use(new MiniEndpointMiddleware({request, fetch, subscribe}));
  }

  fetch = (key: string, componentId: string, query: queryType, mutate: Mutate): Promise.resolve<*> => {
    return this.app.handleChange({
      request: {
        type: 'fetch',
        key,
        query,
        componentId,
      },
      response: {
        mutate,
      },
    });
  }

  subscribe = (key: string, componentId: string, subjectType: SubjectType, observer: rxjs$Observer<*>): Promise.resolve<*> => {
    return this.app.handleChange({
      request: {
        type: 'subscribe',
        key,
        observer,
        componentId,
        subjectType,
      },
    }).then((ctx) => ctx.response.subscription);
  }

  request = (action: MutateAction) => {
    if (action.type !== 'NOOP') {
      const {key} = action.payload;
      return this.app.handleChange({
        request: {
          type: 'write',
          action,
          key,
        },
      });
    }
  }

  deploy = (key?: string, id?: string) => {
    return this.app.handleChange({
      request: {
        type: 'deploy',
        key,
        id,
      },
    });
  }

  reset = (key?: string, id?: string, query?: {[string]: any}) => {
    return this.app.handleChange({
      request: {
        type: 'reset',
        key,
        id,
        query,
        componentId: key,
      },
    });
  }
}
