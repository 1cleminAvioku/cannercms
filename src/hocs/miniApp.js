// @flow

import * as React from 'react';
import {MiniApp} from '../app';
import type {Map} from 'immutable';
import RefId from 'canner-ref-id';

type Props = {
  refId: RefId,
  value: Map<string, any>,
  renderChildren: (childrenProps: Object, deployButtonProps: Object, cancelButtonProps: Object) => React.Node,
  request: RequestDef,
  fetch: FetchDef,
  subscribe: SubscribeDef,
  componentId: string,
  deploy: DeployDef,
  query: QueryDef
};

export default function createWithMiniApp(Com: React.ComponentType<*>) {
  return withMiniApp(Com, MiniApp);
}

export function withMiniApp(Com: React.ComponentType<*>, MiniApp: MiniApp) {
  return class ComponentWithMiniApp extends React.Component<Props> {
    app: MiniApp;

    getProps() {
      return {
        fetch: this.app.fetch,
        subscribe: this.app.subscribe,
        request: this.app.request,
        deploy: this.deploy,
        reset: this.app.reset
      };
    }

    constructor(props: Props) {
      super(props);
      const {request, fetch, subscribe} = props;
      this.app = new MiniApp({
        request,
        fetch,
        subscribe
      });
    }

    componentWillMount() {
      this.fetchData();
    }

    componentWillReceiveProps(nextProps: Props) {
      if (nextProps.refId.toString() !== this.props.refId.toString()) {
        this.fetchData();
      }
    }

    fetchData = () => {
      const {query, componentId, refId} = this.props;
      this.app.fetch(refId.getPathArr()[0], componentId, query);
    }

    deploy = (key?: string, refId?: RefId, callback?: Function): Promise<*> => {
      const {deploy} = this.props;
      const idString = refId && refId.toString();
      return this.app.deploy(key, idString).then(() => {
        return deploy(key, idString).then(callback);
      });
    }

    reset = (key?: string, refId?: RefId, callback?: Function): Promise<*> => {
      const {query} = this.props;
      return this.app.reset(key, refId && refId.toString(), query).then(callback)
    }

    render() {
      const {renderChildren, refId, value} = this.props;
      const key = refId.getPathArr()[0];
      const recordId = value && value.get('_id');
      if (typeof recordId !== 'string') {
        throw new Error(`recordId is not a string, path: ${refId.toString()}`)
      }
      const newRenderChildren = (childrenProps = {}, deployProps = {}, cancelProps = {}) => {
        deployProps.key = deployProps.key || key;
        deployProps.refId = deployProps.refId || new RefId(recordId);
        deployProps.onClick = this.deploy;
        cancelProps.key = cancelProps.key || key;
        cancelProps.refId = cancelProps.refId || new RefId(recordId);
        cancelProps.onClick = this.reset;
        return renderChildren(childrenProps, deployProps, cancelProps);
      };
      return <Com {...this.props} {...this.getProps()} renderChildren={newRenderChildren}/>;
    }
  };
}