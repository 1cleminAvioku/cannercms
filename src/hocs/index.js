import query from './query';
import route from './route';
import cache from './cache';
import request from './request';
import deploy from './deploy';
import id from './id';
import title from './title';
import containerRouter from './containerRouter';
import onDeploy from './onDeploy';
import context from './connectContext';

export default {
  query,
  route,
  cache,
  request,
  deploy,
  // validator,
  id,
  title,
  containerRouter,
  onDeploy,
  context
};