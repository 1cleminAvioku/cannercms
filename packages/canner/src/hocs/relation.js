// @flow

import * as React from 'react';
import {Spin, Icon} from 'antd';
import RefId from 'canner-ref-id';
import Toolbar from './components/toolbar';
import {mapValues} from 'lodash';
import type {HOCProps} from './types';
import {parseConnectionToNormal, getValue, defaultValue} from './utils';
import {withApollo} from 'react-apollo';
import gql from 'graphql-tag';
import {Query} from '../query';
import {List} from 'react-content-loader';

const antIcon = <Icon type="loading" style={{fontSize: 24}} spin />;

type State = {
  originRootValue: any,
  isFetching: boolean,
}

type Props = HOCProps & {
  client: any
}

@withApollo
export default function withQuery(Com: React.ComponentType<*>) {
  // this hoc will fetch data;
  return class ComponentWithQuery extends React.PureComponent<Props, State> {
    query: Query;

    constructor(props: Props) {
      super(props);
      this.state = {
        originRootValue: null,
        isFetching: true
      };
      if (props.relation) {
        this.query = new Query({schema: props.schema});
      }
    }

    componentDidMount() {
      // defaultSort
      const {relation, toolbar} = this.props;
      if (!relation) {
        return;
      }
      if (toolbar && toolbar.async) {
        const args = this.getArgs();
        this.updateQuery([relation.to], {
          ...args,
          first: 10
        });
      } else {
        this.queryData();
      }
      
    }

    UNSAFE_componentWillReceiveProps(props: Props) {
      const {refId, relation} = this.props;
      if (!relation) {
        return;
      }
      if (refId.toString() !== props.refId.toString()) {
        // refetch when route change
        this.queryData(props);
      }
    }

    queryData = (props?: Props): Promise<*> => {
      const {relation, client} = props || this.props;
      if (!relation) {
        return Promise.resolve();
      }
      this.setState({
        isFetching: true,
      });
      const gqlStr = this.query.toGQL(relation.to);
      const variables = this.query.getVairables();
      return client.query({
        query: gql`${gqlStr}`,
        variables
      }).then(({data}) => {
          this.setState({
            originRootValue: data,
            isFetching: false,
          });
        })
        .catch(() => {
          this.setState({
            isFetching: false
          })
        });
    }

    getArgs = () => {
      const {relation} = this.props;
      const queries = this.query.getQueries([relation.to]).args || {pagination: {first: 10}};
      const variables = this.query.getVairables();
      const args = mapValues(queries, v => variables[v.substr(1)]);
      return args;
    }

    updateQuery = (paths: Array<string>, args: Object) => {
      this.query.updateQueries(paths, 'args', args);
      this.queryData();
    }

    render() {
      const {originRootValue, isFetching} = this.state;
      const {toolbar, relation, schema, refId} = this.props;
      if (!relation) {
        return <Com {...this.props}/>;
      }
      console.log(originRootValue);
      if (!originRootValue) {
        return <List style={{maxWidth: 500}} />;
      }
      const value = originRootValue[relation.to];
      const args = this.getArgs();
      const relationValue = value ? removeSelf(value, refId, relation.to) : defaultValue('connection');
      const tb = ({children, ...restProps}) => <Toolbar {...restProps}
        items={schema[relation.to].items.items}
        toolbar={toolbar || {pagination: {type: 'pagination'}}}
        args={args}
        query={this.query}
        keyName={relation.to}
        refId={new RefId(relation.to)}
        originRootValue={originRootValue}
        updateQuery={this.updateQuery}
        parseConnectionToNormal={parseConnectionToNormal}
        getValue={getValue}
        defaultValue={defaultValue}
      >
        <Spin indicator={antIcon} spinning={isFetching}>
          {children}
        </Spin>
      </Toolbar>;
      return <Com {...this.props} Toolbar={tb} relationValue={relationValue}/>;
    }
  };
}

export function removeSelf(value: any, refId: RefId, relationTo: string) {
  const [key, index] = refId.getPathArr().slice(0, 2);
  if (key !== relationTo) {
    return value;
  }
  return {...value, edges: value.edges.filter((v, i) => i !== Number(index))};
}
