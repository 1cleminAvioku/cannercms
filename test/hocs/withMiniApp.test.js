/**
|--------------------------------------------------
| this is just for copy and past
|--------------------------------------------------
*/


import * as React from 'react';
import Enzyme, { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import {withMiniApp} from '../../src/hocs/miniApp';

Enzyme.configure({ adapter: new Adapter() });

describe('hocTemplate', () => {
  let WrapperComponent, props, MockComponent, MiniApp,
  appFetchMock, appDeployMock, appRequestMock, appSubscribeMock, appResetMock;

  beforeEach(() => {
    MockComponent = function MockComponent({renderChildren}) {
      return (<div >{renderChildren && renderChildren()}</div>);
    };
    appFetchMock = jest.fn().mockImplementation(() => Promise.resolve());
    appRequestMock = jest.fn().mockImplementation(() => Promise.resolve());
    appSubscribeMock = jest.fn().mockImplementation(() => Promise.resolve());
    appResetMock = jest.fn().mockImplementation(() => Promise.resolve());
    appDeployMock = jest.fn().mockImplementation(() => Promise.resolve());

    MiniApp = class MiniApp {
      constructor() {}
      deploy = appDeployMock
      fetch = appFetchMock
      subscribe = appSubscribeMock
      reset = appResetMock
      request = appRequestMock
    }

    props = {
      id: 'posts'
    }
    WrapperComponent = withMiniApp(MockComponent, MiniApp);
  });

  it('should render and new an app and fetchData', () => {
    const wrapper = shallow(<WrapperComponent {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.instance().app).toBeInstanceOf(MiniApp);
    expect(appFetchMock).toHaveBeenCalledTimes(1);
  });

  it('should pass context to children', () => {
    const mockDeploy = jest.fn().mockImplementation(() => Promise.resolve());
    const wrapper = shallow(<WrapperComponent {...props} />
    , {
      context: {
        request: () => Promise.resolve(),
        fetch: () => Promise.resolve(),
        subscribe: () => Promise.resolve(),
        deploy: mockDeploy,
        componentId: props.id
      }
    });
    const context = wrapper.instance().getChildContext();
    appFetchMock.mockClear();
    appSubscribeMock.mockClear();
    appRequestMock.mockClear();
    appDeployMock.mockClear();
    appResetMock.mockClear();
    return Promise.all([
      context.fetch(),
      context.subscribe(),
      context.request(),
      context.deploy(),
      context.reset()
    ]).then(() => {
      expect(appFetchMock).toHaveBeenCalledTimes(1);
      expect(appSubscribeMock).toHaveBeenCalledTimes(1);
      expect(appRequestMock).toHaveBeenCalledTimes(1);
      expect(appDeployMock).toHaveBeenCalledTimes(1);
      expect(appResetMock).toHaveBeenCalledTimes(1);
      expect(mockDeploy).toHaveBeenCalledTimes(1);
    });
  });

  it('should edit renderChildren', () => {
    const mockRenderChildren = jest.fn().mockImplementation(() => <div></div>);
    const wrapper = shallow(<WrapperComponent {...props}
      renderChildren={mockRenderChildren}
    />, {
      context: {
        request: () => Promise.resolve(),
        fetch: () => Promise.resolve(),
        subscribe: () => Promise.resolve(),
        deploy: () => Promise.resolve(),
        componentId: props.id
      }
    });
    wrapper.render();
    expect(mockRenderChildren).toHaveBeenCalledWith({}, {
      onClick: wrapper.instance().deploy,
      key: props.id,
      id: undefined
    }, {
      onClick: wrapper.instance().reset,
      key: props.id,
      id: undefined
    })
  });
});