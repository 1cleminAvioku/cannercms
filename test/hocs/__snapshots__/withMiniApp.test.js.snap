// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withRouteMiniApp should render and new an app and fetchData 1`] = `
<MockComponent
  componentId="posts"
  deploy={[Function]}
  fetch={
    [MockFunction] {
      "calls": Array [
        Array [
          "posts",
          "posts",
          undefined,
        ],
      ],
    }
  }
  refId={
    CannerRefId {
      "_internalArr": Array [
        "posts",
      ],
      "child": [Function],
      "getPathArr": [Function],
      "remove": [Function],
      "toString": [Function],
    }
  }
  renderChildren={[Function]}
  request={[MockFunction]}
  reset={[MockFunction]}
  subscribe={[MockFunction]}
  value={
    Immutable.Map {
      "_id": "test",
    }
  }
/>
`;
